lib/rpmrc.cc

```diff
@@ -358,6 +361,105 @@ const char * lookupInDefaultTable(const char * name,
     return name;
 }
 
+static int moveConfigFiles(std::string userdir)
+{
+    fs::path home = std::getenv("HOME");
+    std::error_code ec;
+    std::error_code ec2;
+    fs::path oldmacros = home / ".rpmmacros";
+    fs::path oldrpmrc = home / ".rpmrc";
+    fs::path lockfile = home / ".rpmrc_conversion_lock";
+    fs::path macros_target = "";
+    fs::path rpmrc_target = "";
+
+    if (userdir.substr(0, 2) == "~/")
+	userdir.replace(0, 1, home);
+
+    fs::path userdir_path = userdir;
+    fs::path newmacros = userdir_path / "macros";
+    fs::path newrpmrc = userdir_path / "rpmrc";
+
+    if (!isatty(0)) /* only on interactive runs */
+	return -1;
+
+    if (!fs::is_regular_file(oldmacros) && !fs::is_regular_file(oldrpmrc))
+	return -1;
+
+    int fd = open(lockfile.c_str(), O_CREAT|O_EXCL|O_WRONLY, 0644);
+    if (fd < 0)
+	return -1;
+    close(fd);
+
+    /* recheck after acquiring lock */
+    if ((!fs::is_regular_file(oldmacros) && !fs::is_regular_file(oldrpmrc)) ||
+	fs::exists(userdir_path)) {
+	fs::remove(lockfile, ec);
+	return -1;
+    }
+
+    fs::create_directories(userdir_path, ec);
+    if (ec) goto err2;
+
+    if (fs::is_regular_file(oldmacros)) {
+	fs::copy(oldmacros, newmacros, ec);
+	if (ec) goto err;
+	macros_target = fs::relative(newmacros, home, ec);
+	if (ec) goto err;
+    }
+    if (fs::is_regular_file(oldrpmrc)) {
+	fs::copy(oldrpmrc, newrpmrc, ec);
+	if (ec) goto err;
+	rpmrc_target = fs::relative(newrpmrc, home, ec);
+	if (ec) goto err;
+    }
+
+    if (fs::is_regular_file(oldmacros)) {
+	fs::remove(oldmacros, ec);
+	if (ec) goto undo_remove;
+	fs::create_symlink(macros_target , oldmacros, ec);
+	if (ec) goto undo_remove;
+    }
+    if (fs::is_regular_file(oldrpmrc)) {
+	fs::remove(oldrpmrc, ec);
+	if (ec) goto undo_remove;
+	fs::create_symlink(rpmrc_target , oldrpmrc, ec);
+	if (ec) goto undo_remove;
+    }
+
+    if (fs::is_symlink(oldmacros))
+	rpmlog(RPMLOG_WARNING, "Migrated %s to %s (leaving a symlink)\n",
+	       oldmacros.c_str(), newmacros.c_str());
+    if (fs::is_symlink(oldrpmrc))
+	rpmlog(RPMLOG_WARNING, "Migrated %s to %s (leaving a symlink)\n",
+	       oldrpmrc.c_str(), newrpmrc.c_str());
+
+    fs::remove(lockfile, ec);
+
+    return 0;
+undo_remove:
+    if (fs::is_symlink(oldmacros))
+	fs::remove(oldmacros, ec2);
+    if (fs::is_symlink(oldrpmrc))
+	fs::remove(oldrpmrc, ec2);
+
+    if (fs::is_regular_file(newmacros))
+	fs::copy(oldmacros, newmacros, ec2);
```

Comment:
See above for the copy() grumbling: it's not a good undo when it messes permissions and timestamps.